PyISY/isy.py

```diff
@@ -0,0 +1,248 @@
+    def auto_update(self, val):
+        """Set the auto_update property."""
+        if val and not self.auto_update:
+            # create new event stream socket
+            self._events = EventStream(self, self.conn.connection_info,
+                                       self._on_lost_event_stream)
+        if self._events is not None:
+            self._events.running = val
+
+    def _on_lost_event_stream(self):
+        """Handle lost connection to event stream."""
+        del self._events
+        self._events = None
+
+        if self.auto_reconnect and self._reconnect_thread is None:
+            # attempt to reconnect
+            self._reconnect_thread = Thread(target=self._auto_reconnecter)
+            self._reconnect_thread.daemon = True
+            self._reconnect_thread.start()
+
+    def _auto_reconnecter(self):
+        """Auto-reconnect to the event stream."""
+        while self.auto_reconnect and not self.auto_update:
+            self.log.warning('PyISY attempting stream reconnect.')
+            del self._events
+            self._events = EventStream(self, self.conn.connection_info,
+                                       self._on_lost_event_stream)
+            self._events.running = True
+
+        if not self.auto_update:
+            del self._events
+            self._events = None
+            self.log.warning('PyISY could not reconnect to the event stream.')
+        else:
+            self.log.warning('PyISY reconnected to the event stream.')
+
+        self._reconnect_thread = None
+
+    @staticmethod
+    def _add_node_command(cmd_name, command, value=None):
+        doc = "Send {} command to ISY".format(command)
+        if value is None:
+            def cmd(self, val=None):
+                return self.send_cmd(command, val)
+        else:
+            def cmd(self):
+                return self.send_cmd(command, value)
+            doc = "{} with value {}".format(doc, value)
+
+        cmd.__name__ = cmd_name
+        cmd.__doc__ = doc
+        setattr(Node, cmd_name, cmd)
+
+    @staticmethod
+    def _add_pgrm_command(command, cls):
+        doc = "Send program {} command to ISY".format(command)
+
+        def cmd(self):
+            return self.send_pgrm_cmd(command)
+
+        cmd.__name__ = command
+        cmd.__doc__ = doc
+        setattr(cls, command, cmd)
+
+    def _add_commands(self):
+        """
+        Dynamically add functions for the commands listed.
+
+        The function names are translated according to COMMAND_FRIENDLY_NAME
+        in the constants file (e.g. DOF will be available at self.off())
+
+        Any tuple of a command and a UOM value will add all possible commands
+        from the UOM list in constants.
+
+        FUTURE: Only add the commands needed depending on Node type.
+
+        NOTE: `on` and climate setpoints are special and are still defined
+        explicitly in the Node class.
+        """
+        commands = ['DFON', 'DFOF', 'BRT', 'DIM', 'BEEP',
```

Comment:
The list should only contain one type. That way you don't have to build logic later to handle the fact that you don't know what you're getting out of the list.

```python
command = [('DFON', None), ..., ('SECMD', '84'), ...]
...
for command, values in commands:
```

Trims down a few lines.

Reply:
Updated. Curious what your thoughts are on this in general: I added it to dynamically add the control functions because they were otherwise causing a lot of redundant code.  Eventually, I could see this being used to only add relevant functions depending on Node types (e.g. don't add climate setpoint functions to a light switch)--for now it just eliminates redundant code.

Reply:
I guess I have three questions on that:
1) How much work is it to make this dependent on the node type?
2) Will it break backwards compatibility with ISY firmware 4?
3) Do we care about question 2?

Reply:
1. More than I care to put forth for this update but doable down the road if it was a priority. The logic is built for evaluating types, but it's in Home Assistant's implementation, not this library.
2. Compatibility would be dependent on how you evaluate the types, HA still relies pretty heavily on insteon types as the primary method for sorting, which works for 4 and 5. My changes have expanded sorting for zwave which is also the same, but v5 node servers is a different story.
3. Until v5 has a timeline for exiting beta, besides Polyisy in October, I think v4 still needs to be supported. I've had at least a few people on the HA forum still using v4.

Reply:
So this will work with 4, 5, but probably not node servers? If so, I’m ok with that.